### PR TITLE
Only requeue package revision controller on changes to installed objects

### LINKED ISSUE #1923: Should package revisions watch the objects they control
### What problem are you facing?

I believe the `ProviderRevision` and `ConfigurationRevision` controllers currently do not watch the objects they control (i.e. Deployment, ServiceAccount, CRD, XRD, and Compositions), and thus do not react immediately to reset changes made to those objects. The controllers instead speculatively requeue every 60 seconds.

### How could Crossplane help solve your problem?

The revision controllers could watch the objects they are a controller reference of such that a reconcile would be queued for the revision when any of those objects changed. I believe this would allow us to avoid having the revision controllers speculatively reconcile every 60 seconds.

### PR DESCRIPTION
**Pull request by hasheddan**
### Description of your changes

Updates the package revision controller to only requeue when an installed object changes or when the `____Revision` itself is modified.

Fixes #1923 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

manual testing steps to be added

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: pkg/controller/pkg/revision/reconciler.go [hunk 1]
```diff
@@ -217,6 +224,10 @@ func SetupProviderRevision(mgr ctrl.Manager, l logging.Logger, cache xpkg.Cache,
 	return ctrl.NewControllerManagedBy(mgr).
 		Named(name).
 		For(&v1alpha1.ProviderRevision{}).
+		Watches(&source.Kind{Type: &extv1.CustomResourceDefinition{}}, &handler.EnqueueRequestForOwner{OwnerType: &v1alpha1.ProviderRevision{}}).
+		Watches(&source.Kind{Type: &extv1beta1.CustomResourceDefinition{}}, &handler.EnqueueRequestForOwner{OwnerType: &v1alpha1.ProviderRevision{}}).
```
Comment (negz) [inline pkg/controller/pkg/revision/reconciler.go:228]: Do we need to watch both of these versions? I'm fairly sure that when one changes we'll see an update for the other.

File: pkg/controller/pkg/revision/establisher.go [hunk 2]
```diff
@@ -131,30 +133,42 @@ func (e *APIEstablisher) Establish(ctx context.Context, objs []runtime.Object, p
...
-func (e *APIEstablisher) create(ctx context.Context, obj resource.Object, parent resource.Object, control bool, opts ...client.CreateOption) error {
+func (e *APIEstablisher) create(ctx context.Context, obj runtime.Object, parent resource.Object, control bool, opts ...client.CreateOption) error {
+	o := obj.DeepCopyObject().(resource.Object)
 	ref := meta.AsController(meta.TypedReferenceTo(parent, parent.GetObjectKind().GroupVersionKind()))
 	if !control {
 		ref = meta.AsOwner(meta.TypedReferenceTo(parent, parent.GetObjectKind().GroupVersionKind()))
 	}
 	// Overwrite any owner references on the desired object.
-	obj.SetOwnerReferences([]metav1.OwnerReference{ref})
+	o.SetOwnerReferences([]metav1.OwnerReference{ref})
```
Comment (negz) [inline pkg/controller/pkg/revision/establisher.go:143]: Won't this be a no-op? It looks like we create a deep copy of obj, then change its owner references, then apply obj. Do we need to make the deepcopy, or can we just assert obj to a `resource.Object` (or `metav1.Object`)?

File: pkg/controller/pkg/revision/establisher.go [hunk 3]
```diff
@@ -131,30 +133,42 @@ func (e *APIEstablisher) Establish(ctx context.Context, objs []runtime.Object, p
...
 	return e.client.Create(ctx, obj, opts...)
 }
 
-func (e *APIEstablisher) update(ctx context.Context, current, desired resource.Object, parent resource.Object, control bool, opts ...client.UpdateOption) error {
+func (e *APIEstablisher) update(ctx context.Context, current, desired runtime.Object, parent resource.Object, control bool, opts ...client.UpdateOption) error {
+	c := current.DeepCopyObject().(resource.Object)
+	d := desired.DeepCopyObject().(resource.Object)
```
Comment (negz) [inline pkg/controller/pkg/revision/establisher.go:149]: Ditto the above comment; do we need to deepcopy here?

File: pkg/controller/pkg/revision/establisher.go [hunk 4]
```diff
@@ -131,30 +133,42 @@ func (e *APIEstablisher) Establish(ctx context.Context, objs []runtime.Object, p
...
 	if !control {
-		meta.AddOwnerReference(current, meta.AsOwner(meta.TypedReferenceTo(parent, parent.GetObjectKind().GroupVersionKind())))
-		return e.client.Update(ctx, current, opts...)
+		ownerRef := meta.AsOwner(meta.TypedReferenceTo(parent, parent.GetObjectKind().GroupVersionKind()))
+		for _, ref := range c.GetOwnerReferences() {
+			if ref == ownerRef {
```
Comment (negz) [inline pkg/controller/pkg/revision/establisher.go:153]: Will struct equality work as expected when the struct has pointer fields? (e.g. `ref.Controller`)?

Comment (negz) [reply pkg/controller/pkg/revision/establisher.go:153]: Also, this logic looks pretty similar to what `meta.AddOwnerReference` does. Presumably we could also copy the original, then call `meta.AddOwnerReference`, then use `cmp` to determine whether they changed, but I suspect your current implementation would be more efficient.

File: pkg/controller/pkg/revision/reconciler.go (hunk 1 above)
Comment (hasheddan) [reply pkg/controller/pkg/revision/reconciler.go:228]: You are correct :slightly_smiling_face: 

> // OwnerType is the type of the Owner object to look for in OwnerReferences.  Only Group and Kind are compared.

File: pkg/controller/pkg/revision/establisher.go (hunk 2 above)
Comment (hasheddan) [reply pkg/controller/pkg/revision/establisher.go:143]: > Won't this be a no-op?

This is to guard against a case where a package attempts to install a resource with owner references already present. The reason why I have moved to deep copying in both the create and update methods is that we call them multiple times (once for dry run, once for actual action), and when we modify the owner references on the first pass, they may be inaccurate on the second (e.g. I noticed pointer values being modified, which can be unexpected if you have already stored the object into a different struct, such as `allObjs` here).

File: pkg/controller/pkg/revision/establisher.go (hunk 3 above)
Comment (hasheddan) [reply pkg/controller/pkg/revision/establisher.go:149]: See comment above. The added tests also demonstrate some of this behavior.

File: pkg/controller/pkg/revision/establisher.go (hunk 4 above)
Comment (hasheddan) [reply pkg/controller/pkg/revision/establisher.go:153]: > Will struct equality work as expected when the struct has pointer fields?

This comparison is only used on owners, never controllers, so we don't actually end up comparing any pointers that are not `nil`.

File: pkg/controller/pkg/revision/reconciler.go (hunk 1 above)
Comment (negz) [reply pkg/controller/pkg/revision/reconciler.go:228]: I see! So should we remove `extv1beta1` here?

File: pkg/controller/pkg/revision/establisher.go (hunk 2 above)
Comment (negz) [reply pkg/controller/pkg/revision/establisher.go:143]: I'm still not quite following here; are you intending to be applying the copy (`o`) below, or the original (`obj`)?

Comment (hasheddan) [reply pkg/controller/pkg/revision/establisher.go:143]: Ah yes sorry, I was looking at the comment view here, that omits that line.. will fix

**Comment by hasheddan** [comment]
#1923 is still relevant, but this PR is stale.

**Pull request closed without merging**